cbcontainers/remote_configuration/change_applier.go

```diff
@@ -6,27 +6,75 @@ import (
 )
 
 // ApplyConfigChangeToCR will modify CR according to the values in the configuration change provided
-func ApplyConfigChangeToCR(change models.ConfigurationChange, cr *cbcontainersv1.CBContainersAgent) {
+// If sensorMetadata is provided, specific supported features will be enabled or disabled based on their compatibility with the requested agent version
+func ApplyConfigChangeToCR(change models.ConfigurationChange, cr *cbcontainersv1.CBContainersAgent, sensorMetadata []models.SensorMetadata) {
 	if change.AgentVersion != nil {
 		cr.Spec.Version = *change.AgentVersion
 
-		// We do not set the tag to the version as that would make it harder to upgrade manually
-		// Instead, we reset any "custom" tags, which will fall back to the default (spec.Version)
-		images := []*cbcontainersv1.CBContainersImageSpec{
-			&cr.Spec.Components.Basic.Monitor.Image,
-			&cr.Spec.Components.Basic.Enforcer.Image,
-			&cr.Spec.Components.Basic.StateReporter.Image,
-			&cr.Spec.Components.ClusterScanning.ImageScanningReporter.Image,
-			&cr.Spec.Components.ClusterScanning.ClusterScannerAgent.Image,
-			&cr.Spec.Components.RuntimeProtection.Sensor.Image,
-			&cr.Spec.Components.RuntimeProtection.Resolver.Image,
-		}
-		if cr.Spec.Components.Cndr != nil {
-			images = append(images, &cr.Spec.Components.Cndr.Sensor.Image)
-		}
+		resetImageTagsInCR(cr)
+		toggleFeaturesBasedOnCompatibility(cr, *change.AgentVersion, sensorMetadata)
+	}
+}
 
-		for _, i := range images {
-			i.Tag = ""
+func resetImageTagsInCR(cr *cbcontainersv1.CBContainersAgent) {
+	// We do not set the tag to the version as that would make it harder to upgrade manually
+	// Instead, we reset any "custom" tags, which will fall back to the default (spec.Version)
+	images := []*cbcontainersv1.CBContainersImageSpec{
+		&cr.Spec.Components.Basic.Monitor.Image,
+		&cr.Spec.Components.Basic.Enforcer.Image,
+		&cr.Spec.Components.Basic.StateReporter.Image,
+		&cr.Spec.Components.ClusterScanning.ImageScanningReporter.Image,
+		&cr.Spec.Components.ClusterScanning.ClusterScannerAgent.Image,
+		&cr.Spec.Components.RuntimeProtection.Sensor.Image,
+		&cr.Spec.Components.RuntimeProtection.Resolver.Image,
+	}
+	if cr.Spec.Components.Cndr != nil {
+		images = append(images, &cr.Spec.Components.Cndr.Sensor.Image)
+	}
+
+	for _, i := range images {
+		i.Tag = ""
+	}
+}
+
+func toggleFeaturesBasedOnCompatibility(cr *cbcontainersv1.CBContainersAgent, version string, sensorMetadata []models.SensorMetadata) {
+	var sensorMetadataForVersion *models.SensorMetadata
+	for _, sensor := range sensorMetadata {
+		if sensor.Version == version {
+			sensorMetadataForVersion = &sensor
+			break
 		}
 	}
+	if sensorMetadataForVersion == nil {
```

Comment:
What is expected to happen in such case? I guess this means no sensor version was matched right? Do we want to return an error for that?

Reply:
Initially I thought we should return error as well and abandon the change, but the following convinced me otherwise:
* For customers, the common case is to use the valid versions. We already do a check on the backend when creating the Change - and if the version is semantic, we require that we know about it (so this check here should never be nil). If the version is not semantic, we assume that it is a special version for the customer so we allow the flow to go through - but that should be a rare case for hotfixes or private betas. We don't expect customers to use versions not in the list otherwise. 
* For development flows, this check doesn't make sense as you often have sensor versions that are not in the "main" list and that is OK

So for these cases we assume nothing about the feature support and leave them as-is. Of course, the customer might make an invalid change but we can't verify that.